File: leetcode/303_range_sum_query_immutable.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class NumArray(object):
    def __init__(self, nums):
        current, self.sums = 0, []
        for num in nums:
            current += num
            self.sums.append(current)

    def sumRange(self, i, j):
        if i:
            return self.sums[j] - self.sums[i - 1]
        return self.sums[j]


if __name__ == '__main__':
    obj = NumArray([-2, 0, 3, -5, 2, -1])

    assert 1 == obj.sumRange(0, 2)
    assert -1 == obj.sumRange(2, 5)
    assert -3 == obj.sumRange(0, 5)

File: leetcode/303_range_sum_query_immutable.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <vector>

using namespace std;

class NumArray {
 public:
    explicit NumArray(vector<int> nums) {
        sums_.push_back(0);
        for (auto num : nums) {
            sums_.push_back(sums_.back() + num);
        }
    }

    int sumRange(int i, int j) {
        return sums_[j + 1] - sums_[i];
    }

 private:
    vector<int> sums_;
};

int main() {
    vector<int> nums = {-2, 0, 3, -5, 2, -1};

    auto obj = NumArray(nums);

    assert(1 == obj.sumRange(0, 2));
    assert(-1 == obj.sumRange(2, 5));
    assert(-3 == obj.sumRange(0, 5));
}
